### ISSUE TITLE
Improve logging

### ISSUE THREAD

**Issue by mfeurer** [automl/auto-sklearn#1289]
Currently, configuring the logger is incredibly hard:

* It is unclear which loggers can be configured at all
* There is an example, but it is hard to adapt

It would be great to have an example that demonstrates how to:

* move the output of a single logger to the command line and keep everything else in the log files
* add multiple custom handlers
* describe which loggers exist

**Comment by 0dminnimda** [comment]
> There is an example, but it is hard to adapt

Where? x_x

**Comment by eddiebergman** [comment]
Hi @0dminnimda,

Please see our examples page :)
* https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_debug_logging.html#sphx-glr-examples-40-advanced-example-debug-logging-py

The reason for the complication is due to the distributed processing and dask. Configurations get lost when processes are being spawned and then there's also the issue of writing to the same file. I'm sure there's a better way to fix it but it would take quite some time to learn what that is and then also test it properly. 

Best,
Eddie

**Comment by 0dminnimda** [comment]
Hey @eddiebergman ;]

> Please see our examples page :)
> 
> * https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_debug_logging.html#sphx-glr-examples-40-advanced-example-debug-logging-py

oh my, that's hard to find ... and it seems like doc search it's working right now, hmm

> I'll also be honest, I'm not sure what loggers do exist at the moment such that they can be configured.

probably [tests](https://github.com/automl/auto-sklearn/blob/master/test/test_util/test_logging.py) could give some insight, it's likely not exclusive, but at least something ^_^

**Comment by eddiebergman** [comment]
Thanks for pointing me to those, I had a look a while ago but had forgotten :)

Looking at the tests, it really doesn't cover our main problem which is the process spawning through Dask. As mentioned, it seems that the new processes forget the configuration when the new process is spawned.

The workaround as far as I can tell is to allow the logger instances to be pickled, as this is how Dask transfers obejcts to a new processes. This is not something the default logger objects support, or at least not while also maintaining their configuration.

Having multiple loggers running in parallel also raises issues with multiple writes to a log file. I think this is why there are logged ports used so that they communicate to one process which is then in charge of writing to the file.

I'm not a fan of these ports and pickling loggers as it adds extra params everywhere. It also seems the tests really don't cover these use cases we have.

My idea for a different solution is each process writes to it's own file and the logs get coalesced (sorted by timestamp) at the end. The second part of the idea is that each process would read the config from a file. This eliminates the need for pickling loggers, passing them around and ports. It does however add complexity for the log accumulation step. I've read a bit on it before as this is certainly not a unique problem but never had time to dig into it and apply it.

If you have other suggestions or are willing to have a go with it, I'd be happy to support you along the way.

Best,
Eddie

**Comment by 0dminnimda** [comment]
Yes, it's somewhat a tricky thing trying to log from multiple processes at the same time.
I don't have any expertise with dask, but hopefully it's not that all trick to setup it for our goal.

The approach of writing to several files and joining them won't work as good for stdout/err so it's probably not the best way to go.
Pickling logers seems hacky, setting them in each process form a config is much cleaner.
Port setting is kind of too verbose, if it was automatic then this would probably be the way to do it.

And I can't help but say it's nice that you're trying hard to make it easy and programmer-friendly *wink* to contribute (and it seems to be quite successful) 
Unfortunately I don't have the time for it right now, but I'll try to come back to this in two weeks, will see how everything goes!

**Comment by eddiebergman** [comment]
No problem, thanks for the comments and feedback :) It's good to revisit it once in a while but unfortunatly it's low on the priority list as there hasn't been many issues about it.

If you get a chance to have a look, feel free to comment back here or just start a PR with it but if you can't find the time, no worries

**Comment by 0dminnimda** [comment]
@eddiebergman where should I start?

**Comment by eddiebergman** [comment]
Hello again @0dminnimda :)

The main maintainer is back today so we'll discuss it today during our meeting and then I'll get back to you!

Best,
Eddie

**Comment by mfeurer** [comment]
> I'll also be honest, I'm not sure what loggers do exist at the moment such that they can be configured.

I also don't know. It would be great if we could somehow parse the code for that when building the docs.

> The workaround as far as I can tell is to allow the logger instances to be pickled

Not a good idea: https://stackoverflow.com/questions/24509650/deadlock-with-logging-multiprocess-multithread-python-script

> Having multiple loggers running in parallel also raises issues with multiple writes to a log file. I think this is why there are logged ports used so that they communicate to one process which is then in charge of writing to the file.

Yes, indeed.

To be honest, I don't know whether we can easily work around the logging server and whether writing to a file and the coalescing is a good idea. That would also require a separate process, and would not allow users to easily change timestamps in the logging output.

> The implementation of these parts is mostly already existing, but how to set it up with a file config would need to be changed. This has the benefit that logging is transparent, we do not need to pass around loggers then.

I don't fully understand, what would be the improvement of using a file config here? And do we pass around loggers at all?

**Comment by eddiebergman** [comment]
@mfeurer I would read my latest comment, it agrees with your points there.

| I also don't know. It would be great if we could somehow parse the code for that when building the docs.
* We should use traditional convention for logger names. This way we don't have to explicitly name them anywhere.

In general, I don't think it makes sense for the loggers to come up with their own unique names, the [convention is `getLogger(__name__)`](https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial) in which the logger is associated with the module `autosklearn.x.y.z`. The only reason to break this is if we want two different files to log under the same name, which I don't see a particular reason for. This allows users to configure logs as they would for any other module without having to find explicit names.
This also means they are namespaces so people can configure `autosklearn` and all logs from of the form `autosklearn.x.y.z` will have the same configuration.

> Not a good idea: https://stackoverflow.com/questions/24509650/deadlock-with-logging-multiprocess-multithread-python-script

* I agree and I thought it was already happening by the class `PickleableClientLogger`. Turns out this is just pickles the information required to recreate the logger that knows which port to connect to, not the logger itself. There are also places where we [pass the specific `logger_port`](https://github.com/automl/auto-sklearn/blob/6cc8bb179fcb023d1c341cf33d2958a16a6935be/autosklearn/ensemble_builder.py#L61) which defeats the purpose of the `PicklableLoggerClient` in the first place.

> I don't fully understand, what would be the improvement of using a file config here? And do we pass around loggers at all?
* We don't pass actual loggers around, just the information to recreate one on the right socket. The benefit of having them read from the config file each time is that it's transparent. See the 3rd codeblock of this [section of the docs](https://docs.python.org/3/howto/logging.html#configuring-logging). If we can specify the `logger_port` in the config file then we can just have at the top of each file, it's transparently handled.
```python
# This is just pseduo code, not an actual implementation reference!

# autoskearn.estimators
class AutosklearnEstimator:

    def __init__(config_path: Path | None, ...):
        if config_path is not None:
            config_path = path_to_default_config
         
         logging_config = load(config_path)
         if "port" not in logging_config:
            logger_port = determine_available_port()
            logging_config["port"] = logger_port
            write_to_expected_path(logging_config)
        
# autosklearn.x.y.z
logging.config.fileConfig(expected_path_to_config)
logger = logging.getLogger(__name__)
```

**Comment by somunslotus1** [comment]
I run auto-sklearn on k8s so I want to get the auto-sklearn log on k8s. I follow the doc:https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_debug_logging.html#sphx-glr-examples-40-advanced-example-debug-logging-py, but it does not print the log in the console, so k8s pod log can not catch the logs, how should I config to print the log in the console and logs also written in the file?
